flang/include/flang/Lower/ConvertExpr.h

```diff
@@ -51,6 +52,21 @@ class AbstractConverter;
 class FirOpBuilder;
 class SymMap;
 
+/// The evaluation of some expressions implies a surrounding context. This
+/// context is abstracted by this class.
+class ExpressionContext {
+public:
+  ExpressionContext() = default;
+  ExpressionContext(llvm::ArrayRef<mlir::Value> lcvs)
+      : loopVars{lcvs.begin(), lcvs.end()} {}
+
+  bool inArrayContext() const { return loopVars.size() > 0; }
+  const std::vector<mlir::Value> &getLoopVars() const { return loopVars; }
+
+private:
+  std::vector<mlir::Value> loopVars{};
```

Comment:
What are "loopVars"?  If those are array indexes, wouldn't something like "subscrVars" be better?  Or does this have something to do with implied do loops?